include/bbp/sonata/hdf5_reader.h

```diff
@@ -0,0 +1,197 @@
+///      // calling `readSelection`.
+///      if(selection.empty()) {
+///        return;
+///      }
+///      hdf5_reader.readSelection(dset, selection);
+///    }
+///
+///    void also_wrong(Selection selection) {
+///      // Wrong because `hdf5_reader` is called `selection.ranges().size()`
+///      // number of times. Which could be different on each MPI rank.
+///      for(auto range : selection.ranges()) {
+///        hdf5_reader.readSelection(dset, Selection(std::vector{range}));
+///      }
+///    }
+///
+///    void subtle(Selection selection, bool flag) {
+///      // If the flag can differ between MPI ranks, this is wrong because
+///      // `readSelection` is called with different `dset`s. If the `flag` must
+///      // be the same on all MPI ranks, this is correct. If this happens in
+///      // the libsonata API, then passing the same `flag` on all MPI ranks becomes
+///      // a requirement for the users, when using a collective reader. Example:
+///      //     pop.get_attribute(attr_name, selection)
+///      if(flag) {
+///        hdf5_reader.readSelection(dset1, selection);
+///      } else {
+///        hdf5_reader.readSelection(dset2, selection);
+///      }
+///    }
+///
+///    void correct(Selection selection) {
+///      // Correct because no matter which branch is taken
+///      // `hdf5_reader.readSelection` is called exactly once.
+///      if(selection.size % 2 == 0) {
+///        hdf5_reader.readSelection(dset, selection);
+///      } else {
+///        hdf5_reader.readSelection(dset, {});
+///      }
+///    }
+///
+class Hdf5Reader
+{
+  public:
+    // The issue here is that on a mac `size_t` is different from
+    // `{,u}int{8,16,32,64}_t` but not on the other two OSes.
+    using supported_1D_types = detail::unique_tuple<uint8_t,
+                                                    uint16_t,
+                                                    uint32_t,
+                                                    uint64_t,
+                                                    int8_t,
+                                                    int16_t,
+                                                    int32_t,
+                                                    int64_t,
+                                                    float,
+                                                    double,
+                                                    size_t,
+                                                    std::string>;
+
+    using supported_2D_types = detail::unique_tuple<std::array<uint64_t, 2>>;
+
+    /// Create a valid Hdf5Reader with the default plugin.
+    Hdf5Reader();
+
+    /// Create an Hdf5Reader with a user supplied plugin.
+    Hdf5Reader(std::shared_ptr<Hdf5PluginInterface<supported_1D_types, supported_2D_types>> impl);
+
+    /// Read the selected subset of the one-dimensional array.
+    ///
+    /// Both selections are canonical, i.e. sorted and non-overlapping. The dataset
```

Comment:
I wonder if it's worth making `CanonicalSelection` a type, so there is no chance a non-canoical version is supplied?

Reply:
The difficulty would be functions like:
```
foo(const Selection& selection) {
  if(is_canonical(selection)) {
    // Broken ownership:
    auto sel = CanonicalSelection(std::move(selection.ranges());

    // Feels needlessly expensive, especially if we cross canonical/non-canonical
    // borders a couple of times.
    auto sel = CanonicalSelection(selection.ranges());
    bar(broken_ownership, expensive);
  }
  else {
    // Easy: copy is happening anyway because of the sorting and merging.
    bar(sortAndMerge(selection));
  }
}
```
Meaning I don't think we can do it late easily. If we try to do it early we might get stuck on functions like:
```
??? generateSelection(...) {
  // Depending on the arguments the created selection happens to
  // be canonical, but we can't tell from the arguments. Hence, we don't know
  // the return type.
} 
```

Naturally, we can have a:
```
class CanonicalSelection : public Selection {}
```
and then do things like:
```
auto sel = Selection();
if(is_canonical(sel)) {
  foo(static_cast<const CanonicalSelection&>(sel));
} else {
  foo(sortAndMerge(sel));
}
```
One can do cheap checks rough checks:
```
bool canonical = dynamic_cast<CanonicalSelection const * const>(&sel) != nullptr;
```
allowing one to do expensive sanity check everywhere by short-circuiting with the above trick. Not sure we'd get style point though.

Reply:
Ok, was an idle thought; we can try and bolt it on some other time, and see the ramifications.  Not needed for this PR.  Thanks for the analysis.